src/theory/bags/bags_rewriter.cpp

```diff
@@ -0,0 +1,427 @@
+
+RewriteResponse BagsRewriter::rewriteMakeBag(const TNode& n) const
+{
+  Assert(n.getKind() == MK_BAG);
+  // return emptybag for negative or zero multiplicity
+  if (n[1].isConst() && n[1].getConst<Rational>().sgn() != 1)
+  {
+    // (mkBag x c) = emptybag where c <= 0
+    NodeManager* nm = NodeManager::currentNM();
+    Node emptybag = nm->mkConst(EmptyBag(n.getType()));
+    return RewriteResponse(REWRITE_DONE, emptybag);
+  }
+  return RewriteResponse(REWRITE_DONE, n);
+}
+
+RewriteResponse BagsRewriter::rewriteBagCount(const TNode& n) const
+{
+  Assert(n.getKind() == BAG_COUNT);
+  if (n[1].isConst() && n[1].getKind() == EMPTYBAG)
+  {
+    // (bag.count x emptybag) = 0
+    NodeManager* nm = NodeManager::currentNM();
+    return RewriteResponse(REWRITE_DONE, nm->mkConst(Rational(0)));
+  }
+  if (n[1].getKind() == MK_BAG && n[0] == n[1][0])
+  {
+    // (bag.count x (mkBag x c) = c where c > 0 is a constant
+    return RewriteResponse(REWRITE_DONE, n[1][1]);
+  }
+  return RewriteResponse(REWRITE_DONE, n);
+}
+
+RewriteResponse BagsRewriter::rewriteUnionMax(const TNode& n) const
+{
+  Assert(n.getKind() == UNION_MAX);
+  if (n[1].getKind() == EMPTYBAG || n[0] == n[1])
+  {
+    // (union_max A A) = A
+    // (union_max A emptybag) = A
+    return RewriteResponse(REWRITE_DONE, n[0]);
+  }
+  if (n[0].getKind() == EMPTYBAG)
+  {
+    // (union_max emptybag A) = A
+    return RewriteResponse(REWRITE_DONE, n[1]);
+  }
+  if (n[1].getKind() == UNION_MAX && (n[0] == n[1][0] || n[0] == n[1][1]))
+  {
+    // (union_max A (union_max A B) = (union_max A B)
+    // (union_max A (union_max B A) = (union_max B A)
+    return RewriteResponse(REWRITE_DONE, n[1]);
+  }
+  if (n[0].getKind() == UNION_MAX && (n[0][0] == n[1] || n[0][1] == n[1]))
+  {
+    // (union_max (union_max A B) A) = (union_max A B)
+    // (union_max (union_max B A) A) = (union_max B A)
+    return RewriteResponse(REWRITE_DONE, n[0]);
+  }
+
+  if (n[1].getKind() == UNION_DISJOINT && (n[0] == n[1][0] || n[0] == n[1][1]))
+  {
+    // (union_max A (union_disjoint A B)) = (union_disjoint A B)
+    // (union_max A (union_disjoint B A)) = (union_disjoint B A)
+    return RewriteResponse(REWRITE_DONE, n[1]);
+  }
+  if (n[0].getKind() == UNION_DISJOINT && (n[0][0] == n[1] || n[0][1] == n[1]))
+  {
+    // (union_max (union_disjoint A B) A) = (union_disjoint A B)
+    // (union_max (union_disjoint B A) A) = (union_disjoint B A)
+    return RewriteResponse(REWRITE_DONE, n[0]);
+  }
+  return RewriteResponse(REWRITE_DONE, n);
+}
+
+RewriteResponse BagsRewriter::rewriteUnionDisjoint(const TNode& n) const
+{
+  Assert(n.getKind() == UNION_DISJOINT);
+  if (n[1].getKind() == EMPTYBAG)
+  {
+    // (union_disjoint A emptybag) = A
+    return RewriteResponse(REWRITE_DONE, n[0]);
+  }
+  if (n[0].getKind() == EMPTYBAG)
+  {
+    // (union_disjoint emptybag A) = A
+    return RewriteResponse(REWRITE_DONE, n[1]);
+  }
+  if ((n[0].getKind() == UNION_MAX && n[1].getKind() == INTERSECTION_MIN)
```

Comment:
This rewrite seems fairly specific.  I would hold off on writing anything too specific like this until we have a better idea of whats needed.

Reply:
I agree it is very specific, but I like it :) Do you suggest to remove it?

Reply:
Its fine, lets keep it :)